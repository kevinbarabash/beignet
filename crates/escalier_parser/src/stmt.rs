use escalier_ast::values::*;

use crate::expr::parse_expression;
use crate::parse_error::ParseError;
use crate::pattern::*;
use crate::type_ann::parse_type_ann;
use crate::util::*;

pub fn parse_statement(
    node: &tree_sitter::Node,
    src: &str,
) -> Result<Option<Statement>, ParseError> {
    let kind = match node.kind() {
        "lexical_declaration" => {
            return parse_declaration(node, false, src);
        }
        "class_declaration" => {
            return parse_class_decl(node, src);
        }
        "expression_statement" => {
            let expr = node.named_child(0).unwrap();
            let expr = parse_expression(&expr, src)?;
            StmtKind::ExprStmt(expr)
        }
        "ambient_declaration" => {
            let decl = node.named_child(0).unwrap();
            return parse_declaration(&decl, true, src);
        }
        "type_alias_declaration" => {
            let name = node.child_by_field_name("name").unwrap();
            let id = Ident {
                loc: SourceLocation::from(&name),
                span: name.byte_range(),
                name: text_for_node(&name, src)?,
            };
            let type_ann = node.child_by_field_name("value").unwrap();
            let type_ann = parse_type_ann(&type_ann, src)?;

            let type_params = parse_type_params_for_node(node, src)?;

            StmtKind::TypeDecl(TypeDecl {
                declare: false,
                id,
                type_ann,
                type_params,
            })
        }
        "for_in_statement" => {
            let left_node = node.child_by_field_name("left").unwrap();
            let right_node = node.child_by_field_name("right").unwrap();
            let body_node = node.child_by_field_name("body").unwrap();

            StmtKind::ForStmt(ForStmt {
                pattern: Box::from(parse_pattern(&left_node, src)?),
                expr: Box::from(parse_expression(&right_node, src)?),
                body: parse_block_statement(&body_node, src)?,
            })
        }
        "return_statement" => {
            let arg = match node.named_child(0) {
                Some(arg) => Some(Box::from(parse_expression(&arg, src)?)),
                None => None,
            };

            StmtKind::ReturnStmt(ReturnStmt { arg })
        }
        "comment" => {
            return Ok(None); // ignore comments
        }
        "ERROR" => {
            return Err(ParseError::from(format!(
                "failed to parse: '{}'",
                text_for_node(node, src)?
            )))
        }
        _ => return Err(ParseError::from(format!("unhandled: {:#?}", node))),
    };

    Ok(Some(Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind,
    }))

    // $.export_statement,
    // $.import_statement,
    // $.debugger_statement,
    // $.expression_statement,
    // $.declaration,
    // $.statement_block,

    // $.if_statement,
    // $.switch_statement,
    // $.for_statement,
    // $.for_in_statement,
    // $.while_statement,
    // $.do_statement,
    // $.try_statement,
    // $.with_statement,

    // $.break_statement,
    // $.continue_statement,
    // $.return_statement,
    // $.throw_statement,
    // $.empty_statement,
    // $.labeled_statement
}

fn parse_declaration(
    node: &tree_sitter::Node,
    declare: bool,
    src: &str,
) -> Result<Option<Statement>, ParseError> {
    if node.has_error() {
        // TODO: get actual error node so that we can report where the error is
        return Err(ParseError::from("Error parsing declaration"));
    }

    let decl = node.child_by_field_name("decl").unwrap();
    let rec = node.child_by_field_name("rec").is_some();

    let name = decl.child_by_field_name("name").unwrap();
    let pattern = parse_pattern(&name, src)?;

    let type_ann = if let Some(type_ann) = decl.child_by_field_name("type") {
        Some(parse_type_ann(&type_ann, src)?)
    } else {
        None
    };

    let kind = if rec {
        // `let fib = fix((fib) => (n) => ...)`
        // TODO: Fix always wraps a lambda

        let value = decl.child_by_field_name("value").unwrap();

        let lambda_expr = Expr {
            loc: SourceLocation::from(&decl),
            span: decl.byte_range(),
            kind: ExprKind::Lambda(Lambda {
                params: vec![EFnParam {
                    pat: pattern.clone(),
                    type_ann: type_ann.clone(),
                    optional: false,
                }],
                body: BlockOrExpr::Expr(Box::from(parse_expression(&value, src)?)),
                is_async: false,
                return_type: None,
                type_params: None, // TODO: support type params on VarDecls
            }),
            inferred_type: None,
        };

        let fix = Expr {
            loc: SourceLocation::from(&decl),
            span: decl.byte_range(),
            kind: ExprKind::Fix(Fix {
                expr: Box::from(lambda_expr),
            }),
            inferred_type: None,
        };

        StmtKind::VarDecl(VarDecl {
            pattern,
            type_ann,
            init: Some(Box::from(fix)),
            declare,
        })
    } else {
        let init = if let Some(init) = decl.child_by_field_name("value") {
            Some(Box::from(parse_expression(&init, src)?))
        } else {
            None
        };

        StmtKind::VarDecl(VarDecl {
            pattern,
            type_ann,
            init,
            declare,
        })
    };

    Ok(Some(Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind,
    }))
}

fn parse_class_decl(node: &tree_sitter::Node, src: &str) -> Result<Option<Statement>, ParseError> {
    if node.has_error() {
        // TODO: get actual error node so that we can report where the error is
        return Err(ParseError::from("Error parsing declaration"));
    }

    let name_node = node.child_by_field_name("name").unwrap();
    let body_node = node.child_by_field_name("body").unwrap();

    let mut class_members: Vec<ClassMember> = vec![];
    let mut cursor = body_node.walk();
    for child in body_node.named_children(&mut cursor) {
        let kind = child.kind();

        eprintln!("child.kind() = {kind}");

        match kind {
            "method_definition" => {
                let name_node = child.child_by_field_name("name").unwrap();
                let key = Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                };

                let params = parse_formal_parameters(
                    &child.child_by_field_name("parameters").unwrap(),
                    src,
                )?;
                let body_node = child.child_by_field_name("body").unwrap();
                let body = parse_block_statement(&body_node, src)?;
                let return_type = match child.child_by_field_name("return_type") {
                    Some(type_ann) => Some(parse_type_ann(&type_ann, src)?),
                    None => None,
                };

                let kind = match child.child_by_field_name("kind") {
                    Some(kind) => match text_for_node(&kind, src)?.as_str() {
                        "get" => MethodKind::Getter,
                        "set" => MethodKind::Setter,
                        "*" => todo!(),
                        _ => panic!("Invalid method kind"), // TODO: report an error
                    },
                    None => MethodKind::Method,
                };

                // TODO: add support to the AST for these
                let is_static = child.child_by_field_name("static").is_some();
                let is_mutating = child.child_by_field_name("mut").is_some();

                let type_params = parse_type_params_for_node(&child, src)?;

                if key.name.as_str() == "constructor" {
                    class_members.push(ClassMember::Constructor(Constructor { params, body }));
                } else {
                    class_members.push(ClassMember::Method(ClassMethod {
                        key,
                        kind,
                        lambda: Lambda {
                            params,
                            body: BlockOrExpr::Block(body),
                            is_async: child.child_by_field_name("async").is_some(),
                            return_type,
                            type_params,
                        },
                        is_static,
                        is_mutating,
                    }))
                }
            }
            "public_field_definition" => {
                let name_node = child.child_by_field_name("name").unwrap();
                let key = Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                };

                let value = match child.child_by_field_name("value") {
                    Some(value) => Some(Box::from(parse_expression(&value, src)?)),
                    None => None,
                };
                let type_ann = match child.child_by_field_name("type") {
                    Some(type_ann) => Some(Box::from(parse_type_ann(&type_ann, src)?)),
                    None => None,
                };

                class_members.push(ClassMember::Prop(ClassProp {
                    key,
                    value,
                    type_ann,
                    is_static: child.child_by_field_name("static").is_some(),
                    is_optional: child.child_by_field_name("optional").is_some(),
                    is_mutable: child.child_by_field_name("mut").is_some(),
                }))
            }
            "private_field_definition" => todo!(),
            _ => todo!(),
        }
    }

    let stmt = Statement {
        loc: SourceLocation::from(node),
        span: node.byte_range(),
        kind: StmtKind::ClassDecl(ClassDecl {
            ident: Ident {
                loc: SourceLocation::from(&name_node),
                span: name_node.byte_range(),
                name: text_for_node(&name_node, src)?,
            },
            class: Box::from(Class {
                ident: Ident {
                    loc: SourceLocation::from(&name_node),
                    span: name_node.byte_range(),
                    name: text_for_node(&name_node, src)?,
                },
                body: class_members,
                type_params: parse_type_params_for_node(node, src)?, // TODO
            }),
        }),
    };

    Ok(Some(stmt))
}

pub fn parse_block_statement(node: &tree_sitter::Node, src: &str) -> Result<Block, ParseError> {
    assert_eq!(node.kind(), "statement_block");

    let mut cursor = node.walk();

    let child_count = node.named_child_count();

    let mut stmts: Vec<Statement> = vec![];

    for (i, child) in node.named_children(&mut cursor).into_iter().enumerate() {
        let is_last = i == child_count - 1;
        if is_last {
            // This is the only place where a named `expression` node
            // should exist.  Everywhere else its contents should be
            // inline.
            if child.kind() == "expression" {
                let expr = child.named_child(0).unwrap();
                let expr = parse_expression(&expr, src)?;
                stmts.push(Statement {
                    loc: SourceLocation::from(&child),
                    span: child.byte_range(),
                    kind: StmtKind::ExprStmt(expr),
                })
            } else if let Some(result) = parse_statement(&child, src)? {
                stmts.push(result);
            }
        } else if let Some(result) = parse_statement(&child, src)? {
            stmts.push(result);
        }
    }

    Ok(Block {
        span: node.byte_range(),
        stmts,
    })
}
