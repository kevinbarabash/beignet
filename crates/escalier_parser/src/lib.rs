#[cfg(target_family = "wasm")]
use std::ffi::CString;
use std::os::raw::c_char;

use escalier_ast::types::Type;
use escalier_ast::values::*;

mod expr;
mod literal;
mod parse_error;
mod pattern;
mod stmt;
mod type_ann;
mod util;

pub use crate::parse_error::ParseError;

use crate::stmt::parse_statement;

#[link(wasm_import_module = "my_custom_module")]
extern "C" {
    fn _log(wasm_str: WasmString);
}

#[repr(C)]
pub struct WasmString {
    pub offset: *const c_char,
    pub length: u32,
}

#[cfg(target_family = "wasm")]
unsafe fn string_to_wasm_string(input: &str) -> WasmString {
    let length = input.len() as u32;
    let offset = CString::new(input).unwrap().into_raw();
    WasmString { offset, length }
}

#[cfg(target_family = "wasm")]
fn log(str: &str) {
    unsafe {
        _log(string_to_wasm_string(str));
    }
}

#[cfg(target_family = "unix")]
fn log(str: &str) {
    eprintln!("{}", str);
}

pub fn parse(src: &str) -> Result<Program<Type>, ParseError> {
    let mut parser = tree_sitter::Parser::new();
    parser
        .set_language(tree_sitter_escalier::language())
        .expect("Error loading escalier language");

    log("hello, world!");

    let tree = parser.parse(src, None).unwrap();

    let root = tree.root_node();

    let kind = root.kind();
    if kind == "program" {
        let mut cursor = root.walk();

        let children = root.children(&mut cursor);

        let mut body: Vec<Statement<Type>> = vec![];
        for child in children {
            if let Some(stmt) = parse_statement(&child, src)? {
                body.push(stmt);
            }
        }

        Ok(Program { body })
    } else {
        Err(ParseError::from("not implemented yet"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    pub fn messages(report: &ParseError) -> String {
        report.to_string()
    }

    #[test]
    fn it_works() {
        let src = r#"
        let add = (a, b) => a + b;
        let sub = (a, b) => a - b;
        let sum = add(5, 10);
        "#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn numbers() {
        insta::assert_debug_snapshot!(parse("10;"));
        insta::assert_debug_snapshot!(parse("1.23;"));
        insta::assert_debug_snapshot!(parse("-10;"));
    }

    #[test]
    fn strings() {
        insta::assert_debug_snapshot!(parse(r#""";"#));
        insta::assert_debug_snapshot!(parse(r#""hello";"#));
        insta::assert_debug_snapshot!(parse("\"line 1\\nline 2\\nline 3\";"));
        insta::assert_debug_snapshot!(parse("\"a \\u2212 b\";"));
        insta::assert_debug_snapshot!(parse("\"hello, \\\"world\\\"!\";"));
    }

    #[test]
    fn template_literals() {
        insta::assert_debug_snapshot!(parse("`Hello, world`"));
        insta::assert_debug_snapshot!(parse("`Hello, ${name}`"));
        insta::assert_debug_snapshot!(parse("`(${x}, ${y})`"));
        insta::assert_debug_snapshot!(parse(r#"`Hello, "world"`"#));
        insta::assert_debug_snapshot!(parse("`foo ${`bar ${baz}`}`"));
        insta::assert_debug_snapshot!(parse("`line 1\\nline 2\\nline 3`"));
        insta::assert_debug_snapshot!(parse("`a \\u2212 b`"));
        // insta::assert_debug_snapshot!(parse(r#"if cond { `${foo}` } else { `${bar}` }"#));
    }

    #[test]
    fn tagged_template_literals() {
        insta::assert_debug_snapshot!(parse("sql`SELECT * FROM ${table} WHERE id = ${id}`"));
    }

    #[test]
    #[ignore]
    fn template_literal_with_mismatched_backtick() {
        insta::assert_debug_snapshot!(parse("`foo ${bar`}`"));
    }

    #[test]
    #[ignore]
    fn interpolation_outside_of_template_literal() {
        insta::assert_debug_snapshot!(parse("`foo ${bar}`${baz}`"));
    }

    #[test]
    fn operations() {
        insta::assert_debug_snapshot!(parse("1 + 2 - 3;"));
        insta::assert_debug_snapshot!(parse("x * y / z;"));
        insta::assert_debug_snapshot!(parse("(a + b) * c;"));
        insta::assert_debug_snapshot!(parse("a == b;"));
        insta::assert_debug_snapshot!(parse("a != b;"));
        insta::assert_debug_snapshot!(parse("a > b;"));
        insta::assert_debug_snapshot!(parse("a >= b;"));
        insta::assert_debug_snapshot!(parse("a < b;"));
        insta::assert_debug_snapshot!(parse("a <= b;"));
        insta::assert_debug_snapshot!(parse("let cond = a != b;"));
        insta::assert_debug_snapshot!(parse("-a;"));
        insta::assert_debug_snapshot!(parse("-(a + b);"));
    }

    #[test]
    fn function_definition() {
        insta::assert_debug_snapshot!(parse("(a, b) => c;"));
        insta::assert_debug_snapshot!(parse("() => 10;"));
        insta::assert_debug_snapshot!(parse("(a) => \"hello\";"));
        insta::assert_debug_snapshot!(parse("(a, ...b) => true;"));
        insta::assert_debug_snapshot!(parse("({x, y}) => x + y;"));
        insta::assert_debug_snapshot!(parse("({x: p, y: q}) => p + q;"));
        insta::assert_debug_snapshot!(parse("(a?: boolean, b?) => c;"));
    }

    #[test]
    fn rust_style_functions() {
        insta::assert_debug_snapshot!(parse("let foo = () => { let x = Math.random(); x };"));
        insta::assert_debug_snapshot!(parse("let bar = () => { let x = Math.random(); x; };"));
    }

    #[test]
    fn multiple_rest_params() {
        match parse("(...a, ...b) => true") {
            Ok(_) => panic!("expected test parse() to return an error"),
            Err(report) => {
                assert_eq!(
                    messages(&report),
                    "ParseError: failed to parse: '(...a, ...b) => true'"
                );
            }
        }
    }

    // #[test]
    // #[ignore]
    // fn optional_params_must_appear_last() {
    //     assert_eq!(
    //         parse("(a?, b) => true"),
    //         Err(ParseError::from("optional params must come last")),
    //     );
    // }

    #[test]
    fn async_await() {
        insta::assert_debug_snapshot!(parse("async () => 10;"));
        insta::assert_debug_snapshot!(parse("let foo = async () => { await 10 };"));
        insta::assert_debug_snapshot!(parse("let foo = async () => await a + await b;"));
        insta::assert_debug_snapshot!(parse("let foo = async () => await bar();"));
    }

    #[test]
    fn function_application() {
        insta::assert_debug_snapshot!(parse("foo();"));
        insta::assert_debug_snapshot!(parse("foo(a, b);"));
        insta::assert_debug_snapshot!(parse("foo(10, \"hello\");"));
        insta::assert_debug_snapshot!(parse("f(x)(g(x));"));
        insta::assert_debug_snapshot!(parse("foo(a, ...b);"));
        let src = r#"
        let S = (f) => (g) => (x) => f(x)(g(x));
        let K = (x) => (y) => x;
        let I = S(K)(K);
        "#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn new_expression() {
        insta::assert_debug_snapshot!(parse("new Array();"));
        insta::assert_debug_snapshot!(parse("new Array(1, 2, 3);"));
    }

    #[test]
    fn declarations() {
        insta::assert_debug_snapshot!(parse("let x = 5;"));
        insta::assert_debug_snapshot!(parse("let x = (a, b) => a + b;"));
        insta::assert_debug_snapshot!(parse("let foo = do {let x = 5; x};"));
        // recursive
        insta::assert_debug_snapshot!(parse("let rec f = () => f();"));
        // mutable
        insta::assert_debug_snapshot!(parse(r#"let mut msg: string = "hello, world";"#));
        insta::assert_debug_snapshot!(parse(
            r#"
            let foo = () => {
                let mut msg: string = "hello, world";
                msg
            };
        "#
        ));
    }

    #[test]
    fn assignments() {
        insta::assert_debug_snapshot!(parse("x = 5;"));
        insta::assert_debug_snapshot!(parse("a.b = c;"));
        insta::assert_debug_snapshot!(parse("a[b] = c;"));
        insta::assert_debug_snapshot!(parse(r#"a["b"] = c;"#));
    }

    #[test]
    fn top_level_expressions() {
        insta::assert_debug_snapshot!(parse("a + b;"));
        insta::assert_debug_snapshot!(parse("123;\n\"hello\";"));
    }

    #[test]
    fn if_else() {
        insta::assert_debug_snapshot!(parse("if (true) { 5 } else { 10 };"));
        insta::assert_debug_snapshot!(parse("if (a) { 5 } else if (b) { 10 } else { 20 };"));
    }

    #[test]
    fn objects() {
        insta::assert_debug_snapshot!(parse("{x: 5, y: 10};"));
        insta::assert_debug_snapshot!(parse("let obj = {x, y};"));
        insta::assert_debug_snapshot!(parse("let obj = {a, b, ...others};"));
    }

    #[test]
    fn jsx() {
        insta::assert_debug_snapshot!(parse("<Foo>Hello</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>{bar}</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>Hello {world}!</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>{<Bar>{baz}</Bar>}</Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo></Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo bar={baz} />"));
        insta::assert_debug_snapshot!(parse("<Foo msg=\"hello\" bar={baz}></Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo><Bar>{baz}</Bar></Foo>"));
        insta::assert_debug_snapshot!(parse("<Foo>hello<Bar/>{world}<Baz/></Foo>"));
        insta::assert_debug_snapshot!(parse(
            "let elem = <div point={point} id=\"point\">Hello, {msg}</div>;"
        ));
    }

    // #[test]
    // #[ignore]
    // fn jsx_head_and_tail_must_match() {
    //     assert_eq!(
    //         parse("<Foo>Hello</Bar>"),
    //         Err(ParseError::from("JSX head and tail elements must match")),
    //     );
    // }

    #[test]
    fn type_annotations() {
        insta::assert_debug_snapshot!(parse("let x: number = 5;"));
        insta::assert_debug_snapshot!(parse("let msg: string = \"hello\";"));
        insta::assert_debug_snapshot!(parse("let add = (a: number, b: number): number => a + b;"));
        insta::assert_debug_snapshot!(parse("let p: Point = {x: 5, y: 10};"));
        insta::assert_debug_snapshot!(parse("let FOO: \"foo\" = \"foo\";"));
    }

    #[test]
    fn decls() {
        insta::assert_debug_snapshot!(parse("let x = 5;"));
        insta::assert_debug_snapshot!(parse("   let x = 5;")); // with leading whitespace
        insta::assert_debug_snapshot!(parse("declare let x: number;"));
        insta::assert_debug_snapshot!(parse("declare let foo: Foo<string>;"));
    }

    #[test]
    fn tuples() {
        insta::assert_debug_snapshot!(parse("let x = [];"));
        insta::assert_debug_snapshot!(parse("let x = [1, 2, 3];"));
        insta::assert_debug_snapshot!(parse("let x = [1, [a, b]];"));
        insta::assert_debug_snapshot!(parse("let foo = () => [a, b];"));
    }

    #[test]
    fn member_access() {
        insta::assert_debug_snapshot!(parse("a.b.c;"));
        insta::assert_debug_snapshot!(parse("foo.bar();"));
        insta::assert_debug_snapshot!(parse("p.x * p.x + p.y * p.y;"));
        insta::assert_debug_snapshot!(parse("foo().bar();"));
        insta::assert_debug_snapshot!(parse("arr[0][1];"));
        insta::assert_debug_snapshot!(parse("arr[x](y);"));
        insta::assert_debug_snapshot!(parse("arr[arr.length - 1];"));
        insta::assert_debug_snapshot!(parse("foo[bar[-1]];"));
    }

    #[test]
    fn type_decls() {
        insta::assert_debug_snapshot!(parse("type Num = number;"));
        insta::assert_debug_snapshot!(parse("type Point = {x: number, y: number};"));
        insta::assert_debug_snapshot!(parse("type Foo<T> = {bar: T};"));
        insta::assert_debug_snapshot!(parse("type Foo<T extends string> = {bar: T};"));
        insta::assert_debug_snapshot!(parse(r#"type Foo<T = "foo"> = {bar: T};"#));
        insta::assert_debug_snapshot!(parse(r#"type Foo<T extends string = "foo"> = {bar: T};"#));
        insta::assert_debug_snapshot!(parse("type CoordNames = keyof Point;"));
        insta::assert_debug_snapshot!(parse("type Foo = typeof foo;"));
        insta::assert_debug_snapshot!(parse("type FooBar = typeof foo.bar;"));
        insta::assert_debug_snapshot!(parse(r#"type C = A["b"][C_Key];"#));
        insta::assert_debug_snapshot!(parse("type Array<T> = {[key: number]: T};"));
        insta::assert_debug_snapshot!(parse("type MutPoint = mut {x: number, y: number};"));
        insta::assert_debug_snapshot!(parse("type MutPoint = {mut x: number, mut y: number};"));
    }

    #[test]
    fn blocks() {
        insta::assert_debug_snapshot!(parse("let foo = do {let x = 5; x};"));
        insta::assert_debug_snapshot!(parse("let foo = do {let x = 5; let y = 10; x + y};"));
        insta::assert_debug_snapshot!(parse("do {let x = 5; let y = 10; x + y};"));
        insta::assert_debug_snapshot!(parse(
            "do {let sum = do {let x = 5; let y = 10; x + y}; sum};"
        ));
        insta::assert_debug_snapshot!(parse("let foo = do {let x = 5; console.log(x); x};"));
        insta::assert_debug_snapshot!(parse("let foo = do {console.log(x); x};"));
    }

    #[test]
    fn destructuring() {
        insta::assert_debug_snapshot!(parse("let {x, y} = point;"));
        insta::assert_debug_snapshot!(parse("let {a, b, ...rest} = letters;"));
        insta::assert_debug_snapshot!(parse("let {p0: {x, y}, p1: {x, y}} = line;"));
        insta::assert_debug_snapshot!(parse("let [a, b, ...rest] = letters;"));
        insta::assert_debug_snapshot!(parse("let [foo, ...[bar, ...rest]] = baz;"));
        insta::assert_debug_snapshot!(parse("let foo = ([a, b]) => a;"));
        insta::assert_debug_snapshot!(parse("let foo = ([a, b]: [string, number]) => a;"));
        insta::assert_debug_snapshot!(parse("let foo = ({a, b}) => b;"));
        insta::assert_debug_snapshot!(parse("let foo = ({a, b}: {a: string, b: number}) => b;"));
        insta::assert_debug_snapshot!(parse("let {mut x, y: mut z} = point;"));
        insta::assert_debug_snapshot!(parse("let foo = ({mut x, y: mut z}) => {};"));
        insta::assert_debug_snapshot!(parse("let [a, mut b, ...rest] = letters;"));
        insta::assert_debug_snapshot!(parse("let foo = ([a, mut b, ...rest]) => {};"));
        insta::assert_debug_snapshot!(parse("let {x, mut y = 10} = point;"));
        insta::assert_debug_snapshot!(parse("let [a, mut b = 98, ...rest] = letters;"));
        // TODO: disallowed patterns, e.g. top-level rest, non-top-level type annotations
    }

    #[test]
    fn array_spread() {
        insta::assert_debug_snapshot!(parse("let tuple = [...a, b];"));
        insta::assert_debug_snapshot!(parse("let tuple = [a, ...b];"));
        insta::assert_debug_snapshot!(parse("let tuple = [1, ...[2, 3]];"));
    }

    #[test]
    #[ignore]
    #[should_panic = "Only one rest is allowed in an object pattern"]
    fn multiple_rests_is_invalid() {
        insta::assert_debug_snapshot!(parse("let {z, ...p, ...q} = point"));
    }

    #[test]
    #[ignore]
    #[should_panic = "Rest should come last in object pattern"]
    fn rest_that_isnt_last_is_invalid() {
        insta::assert_debug_snapshot!(parse("let {...p, z} = point"));
    }

    #[test]
    fn types() {
        insta::assert_debug_snapshot!(parse("let get_bar = <T>(foo: Foo<T>) => foo.bar;"));
        insta::assert_debug_snapshot!(parse("declare let get_bar: (foo: Foo) => T;"));
        insta::assert_debug_snapshot!(parse("let str_arr: string[] = [];"));
        insta::assert_debug_snapshot!(parse("let thunk_arr: (() => undefined)[] = [];"));
        insta::assert_debug_snapshot!(parse("let arr: string[] | number[] = [];"));
        insta::assert_debug_snapshot!(parse(
            "declare let add: ((a: number, b: number) => number) & (a: string, b: string) => string;"
        ));
        insta::assert_debug_snapshot!(parse("let nested_arr: string[][] = [];"));
        let src = r#"
        type Event = 
          | {type: "mousedown", x: number, y: number}
          | {type: "keydown", key: string};
        "#;
        insta::assert_debug_snapshot!(parse(src));
        insta::assert_debug_snapshot!(parse("let mut_arr: mut string[] = [];"));
    }

    #[test]
    fn object_types() {
        insta::assert_debug_snapshot!(parse("type Pick<T, K extends keyof T> = {[P in K]: T[P]};"));
        insta::assert_debug_snapshot!(parse("type Foo<T> = {mut [P in keyof T]?: T[P]};"));
        insta::assert_debug_snapshot!(parse("type Bar<T> = {+mut [P in keyof T]+?: T[P]};"));
        insta::assert_debug_snapshot!(parse("type Baz<T> = {-mut [P in keyof T]-?: T[P]};"));
    }

    #[test]
    fn conditional_types() {
        insta::assert_debug_snapshot!(parse(
            r#"type GetTypeName<T extends number | string> = T extends number ? "number" : "string";"#
        ));
        insta::assert_debug_snapshot!(parse(
            "type Flatten<Type> = Type extends Array<infer Item> ? Item : Type;"
        ));
        insta::assert_debug_snapshot!(parse(
            r#"type GetReturnType<Type> = Type extends (...args: never[]) => infer Return
                ? Return
                : never;
        "#
        ));
    }

    #[test]
    fn pattern_matching() {
        insta::assert_debug_snapshot!(parse(
            r#"
            let bar = match (foo) {
                {x, y: b, z: 5, ...rest} -> "object",
                [a, _, ...rest] -> "array",
                "string" -> "string",
                true -> "true",
                false -> "false",
                n -> "variable",
                _ -> "wildcard"
            };
        "#
        ));
        insta::assert_debug_snapshot!(parse(
            r#"
            let bar = match (foo) {
                {a: {b: {c}}} -> "object",
                _ -> "fallthrough"
            };              
            "#
        ));
        insta::assert_debug_snapshot!(parse(
            r#"
            let bar = match (foo) {
                n is number -> "number",
                {a: a is Array} -> "Array",
                _ -> "fallthrough"
            };
            "#
        ));
        insta::assert_debug_snapshot!(parse(
            r#"
            let bar = match (foo) {
                1 -> "one",
                2 -> "two",
                n if (n < 5) -> "few",
                _ -> "many"
            };
            "#
        ))
    }

    #[test]
    fn if_let() {
        insta::assert_debug_snapshot!(parse(
            r#"
            let bar = if (let {x, y: b, ...rest} = foo) {
                "object"
            } else if (let [a, _, ...rest] = foo) {
                "array"
            } else {
                "other"
            };              
        "#
        ));
        insta::assert_debug_snapshot!(parse(
            r#"
            let bar = if (let {x: x is string} = foo) {
                "object"
            } else if (let [a is Array, _, ...rest] = foo) {
                "array"
            } else {
                "other"
            };              
            "#
        ));
    }

    #[test]
    fn classes() {
        insta::assert_debug_snapshot!(parse("class Foo { a: number; }"));
        insta::assert_debug_snapshot!(parse("class Foo { a: number = 5; }"));
        insta::assert_debug_snapshot!(parse("class Foo { a = 5; }"));
        insta::assert_debug_snapshot!(parse("class Foo { static a: number; }"));
        insta::assert_debug_snapshot!(parse("class Foo { foo() {} }"));
        insta::assert_debug_snapshot!(parse("class Foo { foo(): string {} }"));
        insta::assert_debug_snapshot!(parse("class Foo { static foo(): string {} }"));
        insta::assert_debug_snapshot!(parse("class Foo { get foo() {} }"));
        insta::assert_debug_snapshot!(parse("class Foo { set foo(x) {} }"));
        insta::assert_debug_snapshot!(parse("class Foo { constructor(x) {} }"));

        let src = r#"
        class Foo {
            constructor(self, x) {
                self.x = x;
            }
            bar(self, y) {
                self.x + y;
            }
            baz(mut self, x) {
                self.x = x;
            }
        }
        "#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn for_loops() {
        let src = r#"
        let mut sum = 0;
        for (const num of [1, 2, 3]) {
            sum = sum + num;
        }
        "#;
        insta::assert_debug_snapshot!(parse(src));
    }

    // #[test]
    // fn top_level_parse_error() {
    //     let result = parse(
    //         r#"
    //         let obj = {foo: "hello"};
    //         let foo = obj."#,
    //     );
    //     assert_eq!(
    //         result,
    //         Err(ParseError::from("failed to parse: 'let foo = obj.'"))
    //     );
    // }
}
