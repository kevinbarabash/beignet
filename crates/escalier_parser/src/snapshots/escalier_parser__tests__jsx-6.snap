---
source: crates/escalier_parser/src/lib.rs
expression: "parse(\"<Foo bar={baz} />\")"
---
Ok(
    Program {
        body: [
            Statement {
                loc: SourceLocation {
                    start: Position {
                        line: 0,
                        column: 0,
                    },
                    end: Position {
                        line: 0,
                        column: 17,
                    },
                },
                span: 0..17,
                kind: ExprStmt(
                    Expr {
                        loc: SourceLocation {
                            start: Position {
                                line: 0,
                                column: 0,
                            },
                            end: Position {
                                line: 0,
                                column: 17,
                            },
                        },
                        span: 0..17,
                        kind: JSXElement(
                            JSXElement {
                                loc: SourceLocation {
                                    start: Position {
                                        line: 0,
                                        column: 0,
                                    },
                                    end: Position {
                                        line: 0,
                                        column: 17,
                                    },
                                },
                                span: 0..17,
                                name: "Foo",
                                attrs: [
                                    JSXAttr {
                                        loc: SourceLocation {
                                            start: Position {
                                                line: 0,
                                                column: 5,
                                            },
                                            end: Position {
                                                line: 0,
                                                column: 14,
                                            },
                                        },
                                        span: 5..14,
                                        ident: Ident {
                                            loc: SourceLocation {
                                                start: Position {
                                                    line: 0,
                                                    column: 5,
                                                },
                                                end: Position {
                                                    line: 0,
                                                    column: 8,
                                                },
                                            },
                                            span: 5..8,
                                            name: "bar",
                                        },
                                        value: JSXExprContainer(
                                            JSXExprContainer {
                                                loc: SourceLocation {
                                                    start: Position {
                                                        line: 0,
                                                        column: 9,
                                                    },
                                                    end: Position {
                                                        line: 0,
                                                        column: 14,
                                                    },
                                                },
                                                span: 9..14,
                                                expr: Expr {
                                                    loc: SourceLocation {
                                                        start: Position {
                                                            line: 0,
                                                            column: 10,
                                                        },
                                                        end: Position {
                                                            line: 0,
                                                            column: 13,
                                                        },
                                                    },
                                                    span: 10..13,
                                                    kind: Ident(
                                                        Ident {
                                                            loc: SourceLocation {
                                                                start: Position {
                                                                    line: 0,
                                                                    column: 10,
                                                                },
                                                                end: Position {
                                                                    line: 0,
                                                                    column: 13,
                                                                },
                                                            },
                                                            span: 10..13,
                                                            name: "baz",
                                                        },
                                                    ),
                                                    inferred_type: None,
                                                },
                                            },
                                        ),
                                    },
                                ],
                                children: [],
                            },
                        ),
                        inferred_type: None,
                    },
                ),
            },
        ],
    },
)
